cachito/workers/pkg_managers/pip.py

```diff
@@ -0,0 +1,321 @@
+    attribute starting from the top, we start at the bottom. Arguably, starting at the bottom
+    makes more sense, but it should not make any real difference in practice.
+
+    :param ast.AST block_node: Any AST node that has a 'body' attribute
+    :param str attr_name: Name of attribute to search for
+    :param int before_line: Only look for attributes defined before this line
+
+    :rtype: anything that can be expressed as a literal ("primitive" types, collections)
+    :raises AttributeError: If attribute not found or not assigned to a literal expression
+    """
+    if before_line is None:
+        before_line = float("inf")
+    try:
+        return next(
+            ast.literal_eval(node.value)
+            for node in reversed(block_node.body)
+            if node.lineno < before_line and isinstance(node, ast.Assign)
+            for target in node.targets
+            if isinstance(target, ast.Name) and target.id == attr_name
+        )
+    except ValueError:
+        raise AttributeError(f"Attribute {attr_name!r} is not assigned to a literal expression")
+    except StopIteration:
+        raise AttributeError(f"Attribute {attr_name!r} not found")
+
+
+class SetupFile(ABC):
+    """Abstract base class for setup.cfg and setup.py handling."""
+
+    def __init__(self, top_dir, file_name):
+        """
+        Initialize a SetupFile.
+
+        :param str top_dir: Path to root of project directory
+        :param str file_name: Either "setup.cfg" or "setup.py"
+        """
+        self._top_dir = Path(top_dir).resolve()
+        self._path = self._top_dir / file_name
+
+    def exists(self):
+        """Check if file exists."""
+        return self._path.is_file()
+
+    @abstractmethod
+    def get_name(self):
+        """Attempt to determine the package name. Should only be called if file exists."""
+
+    @abstractmethod
+    def get_version(self):
+        """Attempt to determine the package version. Should only be called if file exists."""
+
+
+class SetupCFG(SetupFile):
+    """
+    Parse metadata.name and metadata.version from a setup.cfg file.
+
+    Aims to match setuptools behaviour as closely as possible, but does make
+    some compromises (such as never executing arbitrary Python code).
+    """
+
+    # Valid Python name - any sequence of \w characters that does not start with a number
+    _name_re = re.compile(r"[^\W\d]\w*")
+
+    def __init__(self, top_dir):
+        """
+        Initialize a SetupCFG.
+
+        :param str top_dir: Path to root of project directory
+        """
+        super().__init__(top_dir, "setup.cfg")
+        self._parsed = None
+
+    def get_name(self):
+        """
+        Get metadata.name if present.
+
+        :rtype: str or None
+        """
+        self._parse()
+        name = self._get_option("metadata", "name")
+        if name is None:
+            log.debug("No metadata.name in setup.cfg")
```

Comment:
I suppose I could use `self._path.name` instead, but that will just end up being more typing with not that much upside. I was thinking of somehow injecting the filename into the log messages automatically, but I don't have much experience with more advanced Python logging. Ideally, it would be something like
```python
self.log.debug("Found something in %(filename)s: %s", some_value)  # automatically injects the `filename` arg
```
But so far, I haven't figured out how to do that properly, any ideas?

Worst case scenario, I can fall back to using `self._path.name` everywhere if desirable